File: função de zero/bisseção/bisseção.py
import math
def bisseção(f,epsilon,a,b):
    fa = f(a)
    fb = f(b)
    
    if fa*fb > 0:
        return ValueError()
    
    while (abs(a-b) >= epsilon and abs(fa) >= epsilon and abs(fb) > epsilon):
        x = (a+b)/2
        fx = f(x)
        if fa*fx < 0:
            b = x
            fb = fx
            continue
        if fb*fx < 0:
            a = x
            fa = fx
            continue
        return RuntimeError()
    
    if abs(a-b) < epsilon or abs(fa) < epsilon:
        return a
    if abs(fb) < epsilon:
        return b
    return RuntimeError()

if __name__ == '__main__':
    f = lambda x: math.sin(x)*x**x
    epsilon = 10**-15
    a = 6
    b = 7
    resultado = bisseção(f,epsilon,a,b)
    if isinstance(resultado,BaseException):
        match resultado:
            case ValueError():
                print("Intervalo invalido")
            case _:
                print("Deu error")
    else:
        print(f'A raiz encontrada é {resultado}')
        print(f'F({resultado}) é {f(resultado)}')

File: função de zero/bisseção/bisseção.rs
type Função = fn(f64) -> f64;

enum Erro {
    IntervaloInvalido,
    Desconhecido(String),
}

fn bisseção(f: &Função, epsilon: f64, mut a: f64, mut b: f64) -> Result<f64, Erro> {
    let mut fa = f(a);
    let mut fb = f(b);

    if fa * fb > 0.0 {
        return Err(Erro::IntervaloInvalido);
    }

    while (a - b).abs() >= epsilon && fa.abs() >= epsilon && fb.abs() >= epsilon {
        let x = (a + b) / 2.0;
        let fx = f(x);

        if fx * fa < 0.0 {
            b = x;
            fb = fx;
            continue;
        }

        if fx * fb < 0.0 {
            a = x;
            fa = fx;
            continue;
        }

        return Err(Erro::Desconhecido("Erro dentro do loop".to_string()));
    }

    if (a - b).abs() < epsilon || fa.abs() < epsilon {
        return Ok(a);
    }

    if fb.abs() < epsilon {
        return Ok(b);
    }

    Err(Erro::Desconhecido(
        "Chegou ao final sem encontrar a raiz".to_string(),
    ))
}

fn main() {
    let f: Função = |x| x.sin() * x.powf(x);
    let epsilon = 10.0_f64.powi(-15);
    let a = 6.0;
    let b = 7.0;

    match bisseção(&f, epsilon, a, b) {
        Ok(resultado) => {
            println!("A raiz encontrada é {}", resultado);
            println!("F({}) é {}", resultado, f(resultado))
        }
        Err(Erro::IntervaloInvalido) => println!("O intervalo precisa cruza o eixo x"),
        Err(Erro::Desconhecido(mensagem)) => println!("Deu Erro: {}", mensagem),
    }
}
